src/Nethermind/Nethermind.State/WorldState.cs

```diff
@@ -248,32 +314,110 @@ public Snapshot TakeSnapshot(bool newTransactionStart = false)
 
         public void Restore(Snapshot snapshot)
         {
+            EnsureInScope();
             _persistentStorageProvider.Restore(snapshot.StorageSnapshot.PersistentStorageSnapshot);
             _transientStorageProvider.Restore(snapshot.StorageSnapshot.TransientStorageSnapshot);
             _stateProvider.Restore(snapshot.StateSnapshot);
         }
 
         internal void Restore(int state, int persistentStorage, int transientStorage)
         {
+            EnsureInScope();
             Restore(new Snapshot(state, new Snapshot.Storage(persistentStorage, transientStorage)));
         }
 
         public void SetNonce(Address address, in UInt256 nonce)
         {
+            EnsureInScope();
             _stateProvider.SetNonce(address, nonce);
         }
 
         public void CreateAccountIfNotExists(Address address, in UInt256 balance, in UInt256 nonce = default)
         {
+            EnsureInScope();
             _stateProvider.CreateAccountIfNotExists(address, balance, nonce);
         }
 
-        ArrayPoolList<AddressAsKey>? IWorldState.GetAccountChanges() => _stateProvider.ChangedAddresses();
+        ArrayPoolList<AddressAsKey>? IWorldState.GetAccountChanges()
+        {
+            EnsureInScope();
+            return _stateProvider.ChangedAddresses();
+        }
+
         public void ResetTransient()
         {
+            EnsureInScope();
             _transientStorageProvider.Reset();
         }
 
         PreBlockCaches? IPreBlockCaches.Caches => PreBlockCaches;
+
+        private void EnsureInScope()
+        {
+            if (!_inScope) throw new InvalidOperationException("WorldState used outside of its scope.");
+        }
+
+        private bool _inScope;
+
+        public IDisposable BeginScope(Hash256 stateRoot)
+        {
+            if (_inScope)
+                return new NullScopeGuard();
+
+            // if (_inScope)
+            //     throw new InvalidOperationException("WorldState is already in scope");
+
+            _inScope = true;
+            Reset();
+            StateRoot = stateRoot;
+            return new ScopeGuard(this);
+        }
+
+        public IDisposable BeginScope()
+        {
+            if (_inScope)
+                return new NullScopeGuard();
+
+            // if (_inScope)
+            //     throw new InvalidOperationException("WorldState is already in scope");
+
+            _inScope = true;
+            Reset();
+            return new ScopeGuard(this);
+        }
+
+
+        private void EndScope()
+        {
+            Reset();
+            _inScope = false;
+        }
+
+        private class ScopeGuard(WorldState state) : IDisposable
+        {
+            private bool _disposed;
```

Comment:
I'd make it `int _counter` and do ref counting here. On each `BeginScope` bump up the current, on each dispose -1. When goes to 0, `state.EndScope`.